sixtyfps_compiler/object_tree.rs

```diff
@@ -1612,63 +1612,109 @@ impl Exports {
-                    .and_then(|ident| parser::identifier_text(&ident))
-                    .unwrap_or_else(|| internal_name.clone());
+                    .and_then(|ident| {
+                        parser::identifier_text(&ident).map(|text| (text, ident.clone().into()))
+                    })
+                    .unwrap_or_else(|| {
+                        (internal_name.clone(), export_specifier.ExportIdentifier().clone().into())
+                    });
+
                 NamedExport {
                     internal_name_ident: export_specifier.ExportIdentifier().into(),
                     internal_name,
+                    external_name_ident: name_location,
                     exported_name,
                 }
-            })
-            .collect::<Vec<_>>();
+            },
+        );
 
-        exports.extend(doc.ExportsList().filter_map(|exports| exports.Component()).map(
-            |component| {
+        let exports_it = exports_it.chain(
+            doc.ExportsList().filter_map(|exports| exports.Component()).map(|component| {
+                let name_location: SyntaxNode = component.DeclaredIdentifier().into();
                 let name =
                     parser::identifier_text(&component.DeclaredIdentifier()).unwrap_or_else(|| {
                         debug_assert!(diag.has_error());
                         String::new()
                     });
                 NamedExport {
-                    internal_name_ident: component.DeclaredIdentifier().into(),
+                    internal_name_ident: name_location.clone(),
                     internal_name: name.clone(),
+                    external_name_ident: name_location,
                     exported_name: name,
                 }
-            },
-        ));
-        exports.extend(doc.ExportsList().flat_map(|exports| exports.StructDeclaration()).map(
-            |st| {
+            }),
+        );
+        let exports_it = exports_it.chain(
+            doc.ExportsList().flat_map(|exports| exports.StructDeclaration()).map(|st| {
+                let name_location: SyntaxNode = st.DeclaredIdentifier().into();
                 let name = parser::identifier_text(&st.DeclaredIdentifier()).unwrap_or_else(|| {
                     debug_assert!(diag.has_error());
                     String::new()
                 });
                 NamedExport {
-                    internal_name_ident: st.DeclaredIdentifier().into(),
+                    internal_name_ident: name_location.clone(),
                     internal_name: name.clone(),
+                    external_name_ident: name_location,
                     exported_name: name,
                 }
-            },
-        ));
+            }),
+        );
+
+        struct SeenExport {
+            name_location: SyntaxNode,
+            warned: bool,
+        }
+        let mut seen_exports: HashMap<String, SeenExport> = HashMap::new();
+        let mut export_diagnostics = Vec::new();
```

Comment:
This is needed because `diag` can't be captured mutably by multiple closures. I'd be happy to re-arrange the code if that's desired. This seemed like an okay compromise to me to have the filter condition in one spot.